utils/histogram.hh

```diff
@@ -318,6 +319,89 @@ public:
     }
 };
 
+/*!
+ * \brief A class for a histogram-based summary calculation.
+ *
+ * A summary is a histogram where each bucket holds some quantile.
+ * While a histogram typically holds values from the system start,
+ * a summary is defined over some duration (i.e., latencies in the last 10 seconds).
+ * To calculate a summary, we use two estimated-histograms, calculate their delta, and get the
+ * summary from that delta histogram.
+ *
+ */
+class summary_calculator {
+    std::vector<double> _quantiles = { 0.5, 0.95, 0.99};
+    std::vector<double> _summary = { 0, 0, 0};
+    time_estimated_histogram _previous_histogram;
+    time_estimated_histogram _current_histogram;
+public:
+    /*!
+     * \brief update the summary and histograms
+     *
+     * The update method is called every time tick.
+     * When done, _previous_histogram would equal _current_histogram
+     * and the _summary would contain the current _summary calculation
+     *
+     * The calculation is done in two stages. first, we determine what is
+     * the cutoff for each quantile, for example, assume that there are new 1000
+     * entries and the quantiles are 0.5, 0.95 and 0.99
+     * The cutoffs will be 500, 950, and 990. We reuse the _summary array
+     * to hold these values.
+     *
+     * Second, while coping the _current_histogram to the _previous_histogram,
+     * we collect the diffs. Each time we cross a cutoff value, we update the
+     * _summary with the bucket limit (i.e., the latency value).
+     *
+     * To continue the previous example, if the first 3 diffs had the values:
+     * 10, 300, 200. When reaching the third one, the total diff will be 510,
+     * and we set the summary[0] as the third bucket limit.
+     *
+     */
+    void update() noexcept {
+        auto new_entries = _current_histogram.count() - _previous_histogram.count();
+        if (new_entries <= 0) {
+            return;
+        }
+        for (size_t i = 0; i < _quantiles.size(); i++ ) {
+            _summary[i] = _quantiles[i] * new_entries;
+        }
+        size_t pos = 0;
+        size_t total_diff = 0;
+
+        for (size_t i = 0; i < _current_histogram.size(); i++) {
+            total_diff += _current_histogram[i] - _previous_histogram[i];
+            while (pos < _summary.size() && total_diff >= _summary[pos]) {
+                _summary[pos] = _current_histogram.get_bucket_upper_limit(i);
+                pos++;
+            }
+            _previous_histogram[i] = _current_histogram[i];
+        }
+    }
+
+    const std::vector<double>& quantiles() const noexcept {
+        return _quantiles;
+    }
+
+    void set_quantiles(const std::vector<double>& quantiles) noexcept {
+        _quantiles = quantiles;
```

Comment:
How can this be noexcept?

Should't you reset the other data members?

Reply:
I should probably reset the _summary so it will not return old values, I'll resize it also just in case.